contracts/adapters/IPRebalanceExtension.sol

```diff
@@ -0,0 +1,478 @@
+/*
+    Copyright 2021 Index Coop.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+    SPDX-License-Identifier: Apache License, Version 2.0
+*/
+
+pragma solidity 0.6.10;
+pragma experimental ABIEncoderV2;
+
+import { SafeCast } from "@openzeppelin/contracts/utils/SafeCast.sol";
+import { SafeMath } from "@openzeppelin/contracts/math/SafeMath.sol";
+
+import { AddressArrayUtils } from "../lib/AddressArrayUtils.sol";
+import { BaseExtension } from "../lib/BaseExtension.sol";
+import { GIMExtension } from "./GIMExtension.sol";
+import { IAirdropModule } from "../interfaces/IAirdropModule.sol";
+import { IBaseManager } from "../interfaces/IBaseManager.sol";
+import { IGeneralIndexModule } from "../interfaces/IGeneralIndexModule.sol";
+import { ISetToken } from "../interfaces/ISetToken.sol";
+import { ITransformHelper } from "../interfaces/ITransformHelper.sol";
+import { PreciseUnitMath } from "../lib/PreciseUnitMath.sol";
+
+
+/**
+ * @title IPRebalanceExtension
+ * @author Index Coop
+ *
+ * Manager extension for managing the entire rebalance process for sets that include intrinsic productivity. Utilizes Set Protocol's
+ * GeneralIndexModule, WrapModuleV2, AmmModule and AirdropModule with all actions invoked via the BaseManagerV2 contract. Additionally
+ * uses helper contracts adhering to the ITransformHelper interface to inform the contract how properly interface with Set Protocol to
+ * transform and untransform components and fetch relevent information such as exhcange rates. With this contract, the operator can begin
+ * a rebalance by just supplying components and their target unit allocations mesaured in equivelent amounts of the underlying component.
+ */
+contract IPRebalanceExtension is GIMExtension {
+    using AddressArrayUtils for address[];
+    using PreciseUnitMath for uint256;
+    using SafeCast for int256;
+    using SafeCast for uint256;
+    using SafeMath for uint256;
+
+    /* ============ Structs =========== */
+
+    struct TransformInfo {
+        address underlyingComponent;        // underlying component address
+        uint256 maxTransformSize;           // max transform/untransform size measured in underlying component units
```

Comment:
Why is there a max transform size? Seems like that should be just a wrap and unwrap action.

Reply:
This is because some transformations may have price impact. For example, a transformation could be a single sided deposit into curve using the amm module. To do this safely we need to be able to enforce a max sizes and delays much like the GIM